### PR TITLE
Add comparison operators and bool handling in arithmetic ops

### PR DESCRIPTION
**Pull request by klecki**
#### Why we need this PR?
It adds comparisons and boolean handling to DALI expressions/arithmetic ops.

#### What happened in this PR?
 - Explain solution of the problem, new feature added.
The boolean type was registered for type promotions, support for Constants with booleans and native python `bool` was added.
Implementation for all possible comparisons was added, with safe handling of comparisons between singed and unsigned integers.
Tests were extended and some subset of tests was marked as 'slow' and split into a L1 level test.
 - What was changed, added, removed?
 - What is most important part that reviewers should focus on?
 - Was this PR tested? How?
 - Were docs and examples updated, if necessary?

**JIRA TASK**: [DALI-1178]

### CODE REVIEW

**Comment by klecki** [comment]
!build

File: dali/operators/expressions/arithmetic_test.cc [hunk 1]
```diff
@@ -592,6 +592,59 @@ INSTANTIATE_TEST_SUITE_P(TensorScalarTensorBatch5,
...
+  pipe.RunCPU();
+  pipe.RunGPU();
+  DeviceWorkspace ws;
+  pipe.Outputs(&ws);
+  auto *result0 = ws.OutputRef<CPUBackend>(0).data<int32_t>();
+
+  auto *result1 = ws.OutputRef<GPUBackend>(1).data<int32_t>();
+  vector<int32_t> result1_cpu(batch_size * tensor_elements);
+
+  MemCopy(result1_cpu.data(), result1, batch_size * tensor_elements * sizeof(int));
```
Comment (JanuszL) [inline dali/operators/expressions/arithmetic_test.cc]: Add sync

File: dali/operators/expressions/expression_impl_cpu.h [hunk 2]
```diff
@@ -27,6 +27,29 @@
 
 namespace dali {
 
+template <ArithmeticOp op, typename Result, typename Input0>
```
Comment (JanuszL) [inline dali/operators/expressions/expression_impl_cpu.h]: Why not just input, in other cases we have left and right. And I don't think we will have 3 arguments arithmetic operators.

File: dali/python/nvidia/dali/ops.py [hunk 3]
```diff
@@ -601,11 +622,22 @@ def _choose_device(inputs):
         return "gpu"
     return "cpu"
 
+def _is_boolean_like(input):
```
Comment (JanuszL) [inline dali/python/nvidia/dali/ops.py]: Do we use it anywhere?

File: dali/test/python/test_operator_arithmetic_ops.py [hunk 4]
```diff
@@ -120,68 +146,98 @@ def div_promote(left_type, right_type):
...
+    return result
+
+def float_generator(shape, type, _):
     if (len(shape) == 2):
         return type(np.random.rand(*shape))
     else:
         return type([np.random.rand()])
 
+# Generates inputs of required shapes and types
+# The number of inputs is based on the lenghth of tuple `types`, if types is a single elementm
```
Comment (JanuszL) [inline dali/test/python/test_operator_arithmetic_ops.py]:
```suggestion
# The number of inputs is based on the length of tuple `types`, if types is a single element
```

File: dali/python/nvidia/dali/ops.py (hunk 3 above)
Comment (klecki) [reply dali/python/nvidia/dali/ops.py]: Now it is.

File: dali/test/python/test_operator_arithmetic_ops.py [hunk 5]
```diff
@@ -123,34 +146,50 @@ def div_promote(left_type, right_type):
...
 class ExternalInputIterator(object):
-    def __init__(self, batch_size, shape, types, kinds):
+    def __init__(self, batch_size, shape, types, kinds, disallow_zeros = None):
         try:
             self.length = len(types)
         except TypeError:
             types = (types,)
             kinds = (kinds,)
             self.length = 1
+        if not disallow_zeros:
```
Comment (mzient) [inline dali/test/python/test_operator_arithmetic_ops.py]: Suggestion: how about a sanity check that `disallow_zeros` is `False` when generating `bool`s?

**Comment by klecki** [comment]
!build

**Pull request merged**